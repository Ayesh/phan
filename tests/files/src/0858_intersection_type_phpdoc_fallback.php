<?php
class MyClass858 {
    public function myMethod() {
    }
}

class MyCountableClass extends MyClass858 implements Countable {
    public function count() {
        return 0;
    }
}

/**
 * @param MyClass858&Countable $param
 * @return MyClass858&Countable
 */
function test($param) {
    printf("Count is %d\n", $param->count());
    $param->myMethod();
    $param->missingMethod();
    return $param;
}

// Phan does not currently support real intersection types in phpdoc or inferences.
// This tests that the fallback does something reasonable, and doesn't crash.
function test_infer(MyClass858 $x) {
    if ($x instanceof Countable) {
        test($x); // should not warn
    } elseif ($x instanceof ArrayAccess) {
        test($x); // should warn
    } // TODO: Type combinations should narrow
}

/**
 * @param list<\MyClass858&\Countable> $values
 */
function test_list(array $values) {
    foreach ($values as $value) {
        test($value);
        test_infer($value);
        $value->otherMissingMethod();
    }
}
test(new stdClass());
test(new MyClass858());
test(new MyCountableClass()); // This is a valid cast.
test_list([new stdClass()]);
test_list([new MyClass858()]);
test_list([new MyCountableClass()]); // This is a valid cast.
