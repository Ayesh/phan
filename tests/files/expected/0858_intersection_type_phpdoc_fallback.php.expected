%s:20 PhanUndeclaredMethod Call to undeclared method \MyClass858::missingMethod
%s:30 PhanTypeMismatchArgument Argument 1 ($param) is $x of type \ArrayAccess|\MyClass858|\MyClass858&\ArrayAccess but \test() takes \MyClass858&\Countable defined at %s:17
%s:41 PhanUndeclaredMethod Call to undeclared method \MyClass858::otherMissingMethod
%s:44 PhanTypeMismatchArgument Argument 1 ($param) is new stdClass() of type \stdClass but \test() takes \MyClass858&\Countable defined at %s:17
%s:45 PhanTypeMismatchArgument Argument 1 ($param) is new MyClass858() of type \MyClass858 but \test() takes \MyClass858&\Countable defined at %s:17
%s:47 PhanTypeMismatchArgument Argument 1 ($values) is [new stdClass()] of type array{0:\stdClass} but \test_list() takes list<\MyClass858&\Countable> defined at %s:37
%s:48 PhanTypeMismatchArgument Argument 1 ($values) is [new MyClass858()] of type array{0:\MyClass858} but \test_list() takes list<\MyClass858&\Countable> defined at %s:37
